src/libAtomVM/CMakeLists.txt

```diff
@@ -144,25 +144,35 @@ endif()
 
 if (AVM_DISABLE_SMP)
     target_compile_definitions(libAtomVM PUBLIC AVM_NO_SMP)
-else()
-    include(CheckIncludeFile)
-    CHECK_INCLUDE_FILE(stdatomic.h STDATOMIC_INCLUDE)
-    if(HAVE_PLATFORM_SMP_H)
-        target_compile_definitions(libAtomVM PUBLIC HAVE_PLATFORM_SMP_H)
-    endif()
-    include(CheckCSourceCompiles)
-    check_c_source_compiles("
-        #include <stdatomic.h>
-        int main() {
-            _Static_assert(ATOMIC_POINTER_LOCK_FREE == 2, \"Expected ATOMIC_POINTER_LOCK_FREE to be equal to 2\");
-        }
-    " ATOMIC_POINTER_LOCK_FREE_IS_TWO)
-    if (NOT ATOMIC_POINTER_LOCK_FREE_IS_TWO AND NOT HAVE_PLATFORM_SMP_H)
-        if (NOT STDATOMIC_INCLUDE)
-            message(FATAL_ERROR "stdatomic.h cannot be found, you need to disable SMP on this platform or provide platform_smp.h and define HAVE_PLATFORM_SMP_H")
-        else()
-            message(FATAL_ERROR "Platform doesn't support atomic pointers, you need to disable SMP or provide platform_smp.h and define HAVE_PLATFORM_SMP_H")
-        endif()
+endif()
+if (AVM_DISABLE_TASK_DRIVER)
+    target_compile_definitions(libAtomVM PUBLIC AVM_NO_TASK_DRIVER)
+endif()
+
+if(HAVE_PLATFORM_SMP_H)
+    target_compile_definitions(libAtomVM PUBLIC HAVE_PLATFORM_SMP_H)
+endif()
+if(HAVE_PLATFORM_ATOMIC_H)
+    target_compile_definitions(libAtomVM PUBLIC HAVE_PLATFORM_ATOMIC_H)
+endif()
+
+include(CheckIncludeFile)
+CHECK_INCLUDE_FILE(stdatomic.h STDATOMIC_INCLUDE)
+include(CheckCSourceCompiles)
+check_c_source_compiles("
+    #include <stdatomic.h>
+    int main() {
+        _Static_assert(ATOMIC_POINTER_LOCK_FREE == 2, \"Expected ATOMIC_POINTER_LOCK_FREE to be equal to 2\");
```

Comment:
I don't think we really need to check this, as far as I understand all operations we need are available (and correct) regardless of being lock-free or not:

- https://stackoverflow.com/questions/10106936/are-atomic-variables-lock-free
- https://llvm.org/docs/Atomics.html
- https://gcc.gnu.org/wiki/Atomic/GCCMM/LIbrary

This macro just tells us if we they are implemented on given type using a mutex and some library calls, or using some native instructions, but it shouldn't matter for us.

TL;DR: All atomic types except for std::atomic_flag may be implemented using mutexes or other locking operations, rather than using the lock-free atomic CPU instructions.

Reply:
My understanding of the Atomic keyword and of the atomic functions such as `atomic_compare_exchange_weak` is that if the platform doesn't support lock free atomics (i.e. suitable instructions), the compiler will implement operations through a runtime library. Functions may or may not be part of the runtime libraries. They may use mutexes or critical sections (and mutexes definitely is a problem when this code was meant to be used with interruptions). So we should check and we need to look at every platform and choose the best approach in `platform_atomic.h`.

This code was initially written for Pico and Pico SDK definitely doesn't implement the runtime functions. Code compiles but doesn't link. I'm not sure we should rely on the function names, even if they seem to be the same across compilers, so instead I opted for macros.

Yet, you are right, we probably want to rely on esp-idf implementation. I didn't realize they had one because of the initial Pico experience.

Reply:
I did some further research on this topic, in short:
- RPi2040 SDK makes use of newlib
- newlib provides `newlib/libc/include/stdatomic.h` regardless of atomic helper functions availability. I'm not sure this is the correct way of doing this (from the standard point of view), but we have to deal with this. Maybe they are expecting a custom libatomic should be provided or they didn't care about that.

Anyway, in few words, here we are implementing an heuristic for detecting if native instructions are emitted or a helper function is used.
I think that should be properly documented since it is not really obvious.

There is philosophical question here: should we check this? I mean, I would usually expect that most platforms  with `ATOMIC_POINTER_LOCK_FREE != 2` will provide the right helpers. Of course there will be always "peculiar" implementations around, so providing an easy way for overriding the default and providing working helpers is a good thing (as we are doing now for RaspberryPi).

Last advice, this might be an internal cmake module to make things easier to read.

Reply:
In the microcontroller world I don't believe platforms would usually provide the right helpers. stm32 probably doesn't. Helpers require some kind of OS support (for critical sections or mutexes) which is not built into newlib.